### PR TITLE
New MPI execution engine for -e option.

### PR DESCRIPTION
**Pull request by lyndonhenry**
## Summary of Changes

This pull request adds MPI as an execution engine.  Souffle must be built with `./configre --enable-mpi`, which sets `-DUSE_MPI` as a flag.  Users may then run `souffle -empi` to run Datalog programs with MPI. All MPI code is wrapped in a `#ifdef USE_MPI` so does not affect existing behaviour. Note that this is available for compiled programs only.

There are also a few other changes bundled in also. Likely it would be better to have these as separate pull requests, however doing so would at this point take a considerable amount of time. They are not significant, and do not add new features or introduce substantial behavioural differences.

Files with major changes have been indicated *`in italic`*, while those including only minor changes are `not in italic`.

### `.gitignore`
- add line to disclude generated file

### *`.travis.yml`*
- add new tests for MPI on Linux
- re-order some things alphabetically
- separate docker config from packaging

## *`.travis/Dockerfile`
- new file for using docker tests with travis
- only used by mpi tests at the moment

### *`.travis/run_docker.sh`*
- new script to run souffle in a docker container
- used by MPI tests, but generic enough to be used for more in future
- uses ubuntu bionic base image for souffle
- this image is available in the souffle repo on Dockerhub

### `CMakeLists.txt`
- now includes profiler and unit test files
- includes all files in `src` not previously included or newly added

### *`configure.ac`*
- add `-DNDEBUG` when `--enable-debug` is not provided
- this means `assert` statements are removed in optimised code
- add `--enable-mpi` option, which causes `mpiCC` to be used as the compiler and `-DUSE_MPI` to be set as a flag
- add `Mpi.h` as an include file, for MPI support

### `src/AstClause.cpp`
- replace `ASSERT` macro with `assert`

### `src/AstProgram.cpp`
- remove `Macro.h` as include
- replace `ASSERT` macro with `assert`

### `src/AstRelation.h`
- replace `ASSERT` macro with `assert`

### `src/AstSemanticChecker.cpp`
- replace `ASSERT` macro with `assert`
- disable `-e` option in the presence of records more effectively
- disable `-e` option if mpi is enabled and sqlite is used for IO

### *`src/AstTranslator.cpp`*
- remove `Macro.h` as include
- replace `ASSERT` macro with `assert`
- refactor `makeIODirectives` function
- move `makeRam*` lambdas out of loop
- add `makeRamSend` and `makeRamRecv` for MPI messages
- add `makeRamNotify` and `makeRamWait` to handle the symbol table with MPI enabled
- see `src/RamStatement.h` for more details on these new ram classes
- all other changes are just to make MPI work using these new ram classes

### `src/AstTypeAnalysis.cpp`
- remove `Macro.h` as include
- replace `ASSERT` macro with `assert`

### `src/CompiledSouffle.h`
- add `Mpi.h` as include if MPI is enabled

### `src/FileFormatConverter.h`
- remove this useless file

### `src/Global.cpp`
- remove `Macro.h` as include
- replace macros with standard exceptions

### `src/Global.h`
- remove dumb comment

### `src/IndexSetAnalysis.cpp`
- replace `ASSERT` macro with `assert`

### `src/IndexSetAnalysis.h`
- remove `Macro.h` as include
- replace `ASSERT` macro with `assert`

### `src/InlineRelationsTransformer.cpp`
- return `nullptr` in function returning non-void

### `src/Interpreter.cpp`
- replace `ASSERT` macro with `assert`

### `src/Interpreter.h`
- get rid of unused variable warning

### `src/InterpreterInterface.h`
- replace `ASSERT` macro with `assert`
- change parameter types as per changes to compiler interface, effects nothing at all in the interpreter

### `src/InterpreterRelation.h`
- replace `ASSERT` macro with `assert`

### `src/Macro.h`
- remove this file, macros were always a bad idea

### *`src/Makefile.am`*
- remove `FileFormatConverter.h` from build, as per deletion of that file
- include `Mpi.h` in MPI sources if MPI is enabled
- add MPI sources to build
- add unit test for MPI to build if MPI is enabled

### *`src/Mpi.h`*
- new file supporting MPI in souffle
- too much to explain in notes, so any questions please ask in comments

### `src/PrecedenceGraph.cpp`
- remove `Macro.h` as include
- replace `ASSERT` macro with `assert`
- replace `unsigned`'s with `size_t`'s, where what is being referred to is a size or index
- replace `unsigned`'s with `int`'s where what is being referred to is an identifier of something
- note that the use of `int`'s here makes things more consistent with several other areas of the code base
- in particular, we avoid certain conversions between unsigned and singed types

### `src/PrecedenceGraph.h`
- change numerical types as in `src/PrecedenceGraph.cpp`
- add new member functions to `SCCGraph` class to support MPI in `src/AstTranslator.cpp`
- refactor a few method names to be more informative

### `src/RamCondition.h`
- replace `ASSERT` macro with `assert`

### `src/RamNode.h`
- add `RN_*` statements for new ram classes in `src/RamStatement.h`
- all of this has to do with MPI

### `src/RamOperation.h`
- replace `ASSERT` macro with `assert`

### `src/RamProgram.h`
- replace `ASSERT` macro with `assert`

### *`src/RamStatement.h`*
- replace `ASSERT` macro with `assert`
- fix some signed vs unsigned integer stuff to be consistent with other areas of the code
- add `RamRecv` class, used for reception of an MPI message from a given source stratum
- add `RamSend` class, used for sending of an MPI message to a given destination stratum
- add `RamNotify` class, used to notify the master that output relations are about to be sent and that service of the symbol table is no longer required
- add `RamWait` class, used at the master before output relations are sent to first serve the symbol table, then wait for all notifications to be received from slaves

### `src/RamValue.h`
- replace `ASSERT` macro with `assert`

### `src/RamVisitor.h`
- add `FORWARD` and `LINK` statements for new ram classes in `src/RamStatement.h`
- all of this has to do with MPI

### `src/SouffleInterface.h`
- change unsigned numbers to signed, to avoid conversion between these types

### *`src/SymbolTable.h`*
- add new `handleMpiMessages` member function to handle serving the symbol table to slave nodes
- the master node calls this, while each slave uses member functions of the symbol table as before
- all calls to member functions for lookups and resolutions of symbols are directed to the master node from the slaves

### *`src/Synthesiser.cpp`*
- implementation of visitors and code generation for mpi
- there is one master node
- the master node loads relations, sends them to slaves, serves the symbol table to slaves, and waits for a notification from all slaves that the symbol table is no longer required
- once all notifications have been received, the master sends a message to all the slaves letting them know to send their output relations
- the master then receives all output relations from slaves, and writes them to disk
- each slave waits for relations to be sent from the master and all predecessors, computes the fixpoint, and sends new knowledge on to successor slaves
- the slave then sends a notification to the master, awaits the master's message that notifications have been received from all slaves, and sends its output relations to the master

### `src/TypeSystem.cpp`
- replace `ASSERT` macro with `assert`

### `src/Util.h`
- remove `Macro.h` as an include
- replace `ASSERT` macro with `assert`

### `src/WriteStream.h`
- add line between member functions

### `src/main.cpp`
- include `src/PrecedenceGraph.h` to obtain the number of SCCs
- with MPI enabled, execute compiled programs using `mpiexec` with a number of processes one more than the number of SCCs (i.e. for each SCC/stratum plus the master process)
- don't remove binaries and generated cpp files if debugging is enabled
- add MPI as option for execution engine in command line flags
- replace `ERROR` macro with standard runtime errors
- add MPI `init` and `finalize` calls to start and end of main,

### `src/profile/Cli.h`
- remove `-o` option as per removal of `src/FileFormatConverter.cpp`

### `src/souffle2bdd.cpp`
- replace `ERROR` macro with standard runtime errors

### `src/souffle2lb.cpp`
- replace `ERROR` macro with standard runtime errors

### `src/test/file_format_converter_test.cpp`
- remove test as per removal of `src/FileFormatConverter.cpp`

### `src/test/mpi_test.cpp`
- add unit test for MPI
- it's not great, but sufficient just to test MPI is working

### `src/test/test.h`
- initialize and finalize MPI before and after tests respectively

## `tests/interface/insert_for/driver.cpp`
- run clang format on file

## `tests/interface/insert_print/driver.cpp`
- run clang format on file

## `tests/interface/load_print/driver.cpp`
- run clang format on file

## `tests/interface/signal_error/driver.cpp`
- run clang format on file

### `tests/testsuite.at`
- remove MPI error messages from test output
- sort output files before diff due to non-determinism introduced by MPI

### `utilities/git/hooks/pre-commit.sh`
- taken from `souffle-utilities` repo
- is the only file in that repo
- that repo should be abandoned
- this is a much better place to put souffle utilities
- there is some discussion in the issues backlog about this which justifies it

### CODE REVIEW

**Comment by souffle-lang** [comment]
[64% (+0.13%) vs master 64%](http://plang1.it.usyd.edu.au/jenkins/job/PR-RHEL7-gcc7-Coverage/87/)

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,137 +2,175 @@ CMAKE_MINIMUM_REQUIRED(VERSION 3.6)
...
-        src/Trie.h
-        src/TypeSystem.cpp
-        src/TypeSystem.h
-        src/UnaryFunctorOps.h
-        src/UnionFind.h
-        src/Util.h
-        src/WriteStreamCSV.h
-        src/WriteStream.h
-        src/WriteStreamSQLite.h
+        ./src/CompiledIndexUtils.h
```
Comment (mmcgr) [inline CMakeLists.txt]: This is not friendly to maintain if it changes this much. Does it really help IDEs that use it enough to be worth the effort, and the hassle when it is out of sync with the rest of the source?

**Comment by lyndonhenry** [comment]
Yeah, sorry about merging PRs like that -- it came of laziness and impatience, I will develop more incrementally in future. 

All comments I agree on and have been implemented, with each broken into a separate commit.

File: src/SymbolTable.h [hunk 2]
```diff
@@ -29,13 +35,142 @@
 namespace souffle {
 
 /**
- * @class SymbolTable
+ * SymbolTable
```
Comment (mmcgr) [inline src/SymbolTable.h]: Why remove the @class?

**Comment by souffle-lang** [comment]
[64% (+0.14%) vs master 64%](http://plang1.it.usyd.edu.au/jenkins/job/PR-RHEL7-gcc7-Coverage/90/)

**Pull request merged**